### ISSUE TITLE
Unsafe mechanism for turning a primitive into an enum?

### ISSUE THREAD

**Issue by ltratt** [illicitonion/num_enum#6]
I have a situation where I know I have a numbered `enum` where I know that I have a valid primitive for that enum and I'd rather avoid the cost of the `try_into`. Is it plausible/reasonable to add an `unsafe` method for converting primitives into an `enum`? Maybe `UnsafeFromPrimitive`?

**Comment by illicitonion** [comment]
It sounds like you're looking for `std::mem::transmute`, as in the following:

```rust
#[derive(Debug, PartialEq)]
#[repr(u8)]
enum Number {
    Zero,
    One,
    Two,
}

fn main() {
    let one = 1_u8;

    let enum_one: Number = unsafe { ::std::mem::transmute(one) };

    assert_eq!(enum_one, Number::One)
}
```

In general, this crate exists to avoid needing to mechanically write a bunch of tedious match statements; I suppose the hand-written code the macro would expand for `UnsafeFromPrimitive` would look something like:

```rust
#[derive(Debug, PartialEq)]
#[repr(u8)]
enum Number {
    Zero,
    One,
    Two,
}

impl Number {
    unsafe fn from(num: u8) -> Number {
        ::std::mem::transmute(num)
    }
}

fn main() {
    let one = 1_u8;

    let enum_one = unsafe { Number::from(one) };

    assert_eq!(enum_one, Number::One)
}
```

which has the value of having explicit types rather than needing to perhaps ascribe things for inference, but doesn't feel super tedious to write manually...

So I guess my question comes down to: What do you think deriving `UnsafeFromPrimitive` would save you?

I'd want the function to be `unsafe`, because the caller should be aware that they're taking risks, so it wouldn't save you that. I probably wouldn't want to implement the `From` trait itself, because that implies a degree of safety...

And the compiler gives an error if you try to transmute things of different sizes, so if you change the `repr` size (potentially introducing errors in your unsafe code), the compiler will tell you that you need to fix your types.

So yeah; do you have a use-case that isn't met by `std::mem::transmute`?

**Comment by ltratt** [comment]
I can definitely use `transmute`, but that feels like a sledgehammer to crack a nut, in that it almost entirely bypasses the type system (e.g. it will merrily let me transmute a `usize` to a `u8`). What I think would be useful is something like `unsafe_from` which forces me to use the right type but which doesn't do bounds checking (a bit like `get_unchecked` on a `Vec`). So I can still definitely shoot myself in the foot with it, but it's less unsafe than `transmute`.

**Comment by illicitonion** [comment]
> `transmute` [...] almost entirely bypasses the type system (e.g. it will merrily let me transmute a `usize` to a `u8`).

That's not the case; `transmute` will error if the sizes of types aren't the same:
```
error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
 --> src/main.rs:3:9
  |
3 |         std::mem::transmute::<usize, u8>(1_usize);
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: source type: `usize` (64 bits)
  = note: target type: `u8` (8 bits)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0512`.
error: Could not compile `rustplay`.

To learn more, run the command again with --verbose.
```

so you can probably transmute a `u64` and a `usize`, but you can't pass around arbitrary types.

It sounds like you want explicit, rather than inferred, types on the `transmute` call, and that the boilerplate you're looking to save is some combination of writing the turbofish, and having to update the type in the turbofish if the `repr` of the `enum` changes. I think I buy that that's worthwhile.

How does #7 look to you?

**Comment by ltratt** [comment]
You are quite right about `transmute` and types -- mea culpa! That said, #7 looks interesting, so I'll comment there.

**Comment by illicitonion** [comment]
Released as 0.3.1 :) Thanks for the suggestion!
